### ISSUE TITLE
'channels_first' image data format is not fully supported

### ISSUE THREAD

**Issue by tiferet** [keras-team/keras#9670]
https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/engine/training.py#L1460

Using TF backend, I tried switching to `'channels_first'` format, and encountered multiple bugs. For example, when using `sparse_categorical_crossentropy`, [this code](https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/engine/training.py#L1460) assumes `'channels_last'` format.

If that line is changed to correctly support `'channels_first'`, sample weights cannot be used. If the sample weight array is given shape `(batch_size, data_size)`, [this](https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/engine/training.py#L509) error is raised. On the other hand, if it is given shape `(batch_size, 1, data_size)`, [this](https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/engine/training.py#L486) error is raised.

If converting to one-hot encoding instead of using `sparse_categorical_crossentropy`, the function [`to_categorical`](https://github.com/keras-team/keras/blob/068a680480ebf27ecbe57f1406d60157b7d2df38/keras/utils/np_utils.py#L32) also assumes `'channels_last'` format, and puts the classes last.

Even if that is fixed, sample weights still cannot be used.

In short, `'channels_first'` image data format seems not to be fully supported, and many places in the codebase seem to assume `'channels_last'` format.

**Comment by tiferet** [comment]
Some additional places I've run into that assume `'channels_last'` are [this line](https://github.com/keras-team/keras/blob/be112e1f48bbdf24d0a714b9cafc3fc9f43c7738/keras/backend/tensorflow_backend.py#L3041) and [this line](https://github.com/keras-team/keras/blob/be112e1f48bbdf24d0a714b9cafc3fc9f43c7738/keras/backend/tensorflow_backend.py#L3048) in `tensorflow_backend` `sparse_categorical_crossentropy`.

**Comment by ghumpire** [comment]
I also found a similar problem when using `'channels_first'`. For instance, `BatchNormalization` and `Concatenate` need `axis=1`, they assume `'channels_last'` by default.

**Comment by al-rigazzi** [comment]
May I ask if this problem has been fixed? In what version?

**Comment by mmt-lending-tech** [comment]
I found the same problem while working with `ImageDataGenerator`, 
as `array_to_img` fails with `ValueError('Unsupported channel number: %s' % (x.shape[2],))`
When can we expect this issue to be fixed.
